File: galite-core/src/main/java/org/kopi/vkopi/lib/ui/swing/form/DObjectField.java
/*
 * Copyright (c) 1990-2022 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: DObjectField.java 34961 2016-11-04 17:20:49Z hacheni $
 */

package org.kopi.vkopi.lib.ui.swing.form;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.InputEvent;
import java.awt.event.KeyEvent;

import javax.swing.AbstractAction;
import javax.swing.JComponent;
import javax.swing.KeyStroke;

import org.kopi.galite.visual.form.VField;
import org.kopi.galite.visual.form.VFieldUI;
import org.kopi.galite.visual.visual.Action;

/**
 * DImageField is a panel composed in a Image field and a label behind
 */
public abstract class DObjectField extends DField {

    // ----------------------------------------------------------------------
    // CONSTRUCTION
    // ----------------------------------------------------------------------

    /**
     * Constructor
     *
     * @param    model        the model for this text field
     * @param    label        The label that describe this field
     * @param    options        The possible options (NO EDIT, NO ECHO)
     */
    public DObjectField(VFieldUI model,
                        DLabel label,
                        int align,
                        int options,
                        boolean detail) {
        super(model, label, align, options, detail);
        addNavigationKey();
        setRequestFocusEnabled(true);
        setFocusCycleRoot(true);
    }

    // ----------------------------------------------------------------------
    // KEY EVENTS
    // ----------------------------------------------------------------------

    /**
     *
     */
    /*package*/ void addNavigationKey() {
        addKey(KEY_TAB, KeyEvent.VK_ENTER, 0);
        addKey(KEY_TAB, KeyEvent.VK_TAB, 0);
        addKey(KEY_STAB, KeyEvent.VK_TAB, InputEvent.SHIFT_MASK);
        addKey(KEY_BLOCK, KeyEvent.VK_ENTER, InputEvent.SHIFT_MASK);
        addKey(KEY_REC_UP, KeyEvent.VK_PAGE_UP, 0);
        addKey(KEY_REC_DOWN, KeyEvent.VK_PAGE_DOWN, 0);
        addKey(KEY_REC_FIRST, KeyEvent.VK_HOME, 0);
        addKey(KEY_REC_LAST, KeyEvent.VK_END, 0);
        addKey(KEY_STAB, KeyEvent.VK_LEFT, InputEvent.CTRL_MASK);
        addKey(KEY_TAB, KeyEvent.VK_RIGHT, InputEvent.CTRL_MASK);
        addKey(KEY_REC_UP, KeyEvent.VK_UP, InputEvent.CTRL_MASK);
        addKey(KEY_REC_DOWN, KeyEvent.VK_DOWN, InputEvent.CTRL_MASK);
    }

    /**
     *
     */
    private void addKey(int code, int key, int mod) {
        KeyStroke keyStroke = KeyStroke.getKeyStroke(key, mod);
        registerKeyboardAction((ActionListener) new Navigator(code),
                null,
                keyStroke,
                JComponent.WHEN_FOCUSED);
    }

    // ----------------------------------------------------------------------
    // UI MANAGEMENT
    // ----------------------------------------------------------------------

    void enter(boolean refresh) {
        super.enter(refresh);
        requestFocusInWindow();
        //    Done in DField
        //    fireFocusHasChanged();
        setBlink(true);
    }

    void leave() {
        setBlink(false);
        super.leave();
    }

    // ----------------------------------------------------------------------
    // FOCUS
    // ----------------------------------------------------------------------

    /**
     *
     */
    static class Navigator extends AbstractAction {

        Navigator(int keyCode) {
            super("navigation-key");
            this.keyCode = keyCode;
        }

        public void actionPerformed(final ActionEvent e) {
            Action action = null;
            if (e.getSource() == null) {
                return;
            }
            final VField field = ((DField) e.getSource()).getModel();
            if (field == null ||
                    field.getForm() == null ||
                    field.getForm().getActiveBlock() == null ||
                    (field.getForm().getActiveBlock().getActiveField() != field &&
                            keyCode != KEY_BLOCK)) {
                return;
            }
            switch (keyCode) {
                case KEY_TAB:
                    action = new Action("key: next field") {
                        public void execute() {
                            field.getForm().getActiveBlock().gotoNextField();
                        }
                    };
                    break;
                case KEY_STAB:
                    action = new Action("key: previous field") {
                        public void execute() {
                            field.getForm().getActiveBlock().gotoPrevField();
                        }
                    };
                    break;
                case KEY_BLOCK:
                    action = new Action("key: next block") {
                        public void execute() {
                            field.getForm().gotoNextBlock();
                        }
                    };
                    break;
                case KEY_REC_UP:
                    action = new Action("key: previous record") {
                        public void execute() {
                            field.getForm().getActiveBlock().gotoPrevRecord();
                        }
                    };
                    break;
                case KEY_REC_DOWN:
                    action = new Action("key: next record") {
                        public void execute() {
                            field.getForm().getActiveBlock().gotoNextRecord();
                        }
                    };
                    break;
                case KEY_REC_FIRST:
                    action = new Action("key: first record") {
                        public void execute() {
                            field.getForm().getActiveBlock().gotoFirstRecord();
                        }
                    };
                    break;
                case KEY_REC_LAST:
                    action = new Action("key: last record") {
                        public void execute() {
                            field.getForm().getActiveBlock().gotoLastRecord();
                        }
                    };
                    break;
            }

            //      field.getForm().performAction(action, false);
            // REPLACED BY:
            field.getForm().performAsyncAction(action);
        }

        private int keyCode;
    }

    // ----------------------------------------------------------------------
    // DATA MEMBERS
    // ----------------------------------------------------------------------

    private static final int KEY_TAB = 0;
    private static final int KEY_STAB = 1;
    private static final int KEY_REC_UP = 2;
    private static final int KEY_REC_DOWN = 3;
    private static final int KEY_REC_FIRST = 4;
    private static final int KEY_REC_LAST = 5;
    private static final int KEY_BLOCK = 6;
    /**
     * Comment for <code>serialVersionUID</code>
     */
    private static final long serialVersionUID = -4243258834913065429L;
}

File: galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/form/DObjectField.kt
/*
 * Copyright (c) 2013-2021 kopiLeft Services SARL, Tunis TN
 * Copyright (c) 1990-2021 kopiRight Managed Solutions GmbH, Wien AT
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.kopi.galite.visual.ui.vaadin.form

import org.kopi.galite.visual.form.VFieldUI
import org.kopi.galite.visual.ui.vaadin.field.ObjectFieldListener
import org.kopi.galite.visual.visual.Action

/**
 * The `DObjectField` is the vaadin implementation
 * of the object field.
 *
 * @param model The row controller.
 * @param label The field label.
 * @param align The field alignment.
 * @param options The field options.
 * @param detail Does the field belongs to the detail view ?
 */
abstract class DObjectField(model: VFieldUI,
                            label: DLabel?,
                            align: Int,
                            options: Int,
                            detail: Boolean) : DField(model, label, align, options, detail), ObjectFieldListener {

  // --------------------------------------------------
  // UI MANAGEMENT
  // --------------------------------------------------
  override fun enter(refresh: Boolean) {
    super.enter(refresh)
    if (blinkOnFocus()) {
      setBlink(true)
    }
  }

  override fun leave() {
    if (blinkOnFocus()) {
      setBlink(false)
    }
    super.leave()
  }

  //---------------------------------------------------
  // OBJECT FIELD LISTENER IMPLEMENTATION
  //---------------------------------------------------
  override fun gotoNextField() {
    performAction(object : Action("key: next field") {
      override fun execute() {
        getModel().getForm().getActiveBlock()!!.gotoNextField()
      }
    })
  }

  override fun gotoPrevField() {
    performAction(object : Action("key: previous field") {
      override fun execute() {
        getModel().getForm().getActiveBlock()!!.gotoPrevField()
      }
    })
  }

  override fun gotoNextBlock() {
    performAction(object : Action("key: next block") {
      override fun execute() {
        getModel().getForm().gotoNextBlock()
      }
    })
  }

  override fun gotoPrevRecord() {
    performAction(object : Action("key: previous record") {
      override fun execute() {
        getModel().getForm().getActiveBlock()!!.gotoPrevRecord()
      }
    })
  }

  override fun gotoNextRecord() {
    performAction(object : Action("key: next record") {
      override fun execute() {
        getModel().getForm().getActiveBlock()!!.gotoNextRecord()
      }
    })
  }

  override fun gotoFirstRecord() {
    performAction(object : Action("key: first record") {
      override fun execute() {
        getModel().getForm().getActiveBlock()!!.gotoFirstRecord()
      }
    })
  }

  override fun gotoLastRecord() {
    performAction(object : Action("key: last record") {
      override fun execute() {
        getModel().getForm().getActiveBlock()!!.gotoLastRecord()
      }
    })
  }

  /**
   * Executes the given action in the event dispatch handler.
   * @param action The action to be executed.
   */
  protected fun performAction(action: Action?) {
    if (action != null && model != null) {
      getModel().getForm().performAsyncAction(action)
    }
  }

  /**
   * Returns true if the field should be blinked when a focus event is triggered.
   * @return True if the field should be blinked when a focus event is triggered.
   */
  protected open fun blinkOnFocus(): Boolean {
    return true
  }

  override fun valueChanged() {
    // Nothing to do
  }
}
